File: 22-generate-parentheses/generate-parentheses.go
func generateParenthesis(n int) []string {
    ans := make([]string, 0)
    var dfs func(int, int, string)
    dfs = func(l, r int, p string) {
        if l == r && l + r == 2 * n {
            ans = append(ans, p)
        }

        if l < n {
            dfs(l + 1, r, p + "(")
        }

        if r < l {
            dfs(l, r + 1, p + ")")
        }
    }
    dfs(0, 0, "")
    return ans
}

File: 22-generate-parentheses/generate-parentheses.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<string> ans;
        dfs(0, 0, "", n, ans);
        return ans;
    }

    void dfs(int l, int r, string p, int n, vector<string>& ans) {
        if (l == r && l + r == n * 2) {
            ans.push_back(p);
            return;
        }

        if (l < n) {
            dfs(l + 1, r, p + "(", n, ans);
        }

        if (r < l) {
            dfs(l, r + 1, p + ")", n, ans);
        }
    }
};